invocation.md

```diff
@@ -0,0 +1,1170 @@
+
+| UCAN Field | Delegation                             | Task                            |
+| ---------- | -------------------------------------- | ------------------------------- |
+| `iss`      | Delegator: transfer authority (active) | Invoker: request task (active)  |
+| `aud`      | Delegate: gain authority (passive)     | Executor: perform task (active) |
+
+### 2.1.1 Invoker
+
+The invoker signals to the executor that a task associated with a UCAN SHOULD be performed.
+
+The invoker MUST be the UCAN delegator. Their DID MUST be authenticated in the `iss` field of the contained UCAN.
+
+### 2.1.2 Executor
+
+The executor is directed to perform some task described in the UCAN by the invoker.
+
+The executor MUST be the UCAN delegate. Their DID MUST be set the in `aud` field of the contained UCAN.
+
+## 2.2 Components
+
+```mermaid
+flowchart LR
+ U(UCAN) -- Auth --> I
+ I(Invocation) -- Output --> O(Result)
+ O -- Consume --> P(Promise)
+ O -- Attest --> R(Receipt)
+ R -- Short-circuit --> M(Memoization)
+```
+
+### 2.2.1 Invocation
+
+An [Invocation] is like a cryptographically signed function application, a request to perform some action on a resource with specific input. Invocation MAY have optional metadata that is not used to describe the meaning of the computation or effect to be run. Executor MUST produce same result regardless of the metadata.
+
+### 2.2.2 Receipt
+
+A [Receipt] describes the output of an invocation. It is referenced either by the invocation itself or an IPLD Link of the invocation.
+
+### 2.2.3 Promise
+
+A [promise] is a reference to expected result of the invocation receipt.
+
+## 2.3 IPLD Schema
+
+```ipldsch
+type Invocation<In> struct {
+  v      SemVer
+
+  iss    Principal
+  aud    Principal
+
+  with   URI
+  do     Ability
+  input  In (implicit {})
+
+  meta   {String : Any} (implicit {})
+
+  prf    [&UCAN]
+  nnc    optional String
+  nbf    optional Int
+
+  s      Varsig
+}
+
+type Receipt<In, Out> struct {
+  job     &Invocation<In>
+
+  # output of the invocation
+  out     Out
+
+  # Related receipts
+  origin optional &Receipt<In, Any>
+
+  # All the other metadata
+  meta    { String: Any }
+
+  # Principal that issued this receipt
+  iss     Principal
+
+  # Signature from the `iss`.
+  s       Varsig
+
+  # Proof that `iss` was authorized to by invocation `aud` to issue
+  # receipts. Can be omitted if `job.aud === this.iss`
+  prf     [&UCAN] implicit ([])
+}
+
+# Promise is a way to reference data from the output of the invocation
+type Promise struct {
+  Await    "<-"
+} representation keyed
+
+
+type Await union {
+  # Invocation reference
+  | &Invocation<Any>
+  # Inline invocation
+  | Invocation<Any>
+  # Specific invocation output
+  | ResultSelector
+} representation kinded
+
+type ResultSelector struct {
+  job   InvocationReference
+  at    Selector
+} representation tuple
+
+type Selector union {
+  | Key     String
+  | Path    [String]
+} kinded
```

Comment:
We started down this path in the UCAN Invocation spec, but reverted it because of the complexity. Going down this path, we really quickly ran into edge cases, exploring using jsonpath, and so on. We decided instead to just use a step in the workflow that's not baked into the core spec instead (i.e. we can use Wasm or a task type to do e.g. data shaping).

Reply:
I think there is some middle-ground between run WASM and just select ok / error, which is what I was trying to do here. JSONPath is pretty complex and I'm not surprised it did not pan out.

While I still kind of think that array of keys to traverse into is pragmatic and will get you far enough, I'm happy to let it go and maybe do it at other layer.